prow/config/agent.go

```diff
@@ -91,41 +99,46 @@ func (ca *Agent) Start(prowConfig, jobConfig string) error {
 	return nil
 }
 
-// Delta represents the before and after states of a Config change detected by the Agent.
-type Delta struct {
-	Before, After Config
-}
-
 // Subscribe registers the channel for messages on config reload.
 // The caller can expect a copy of the previous and current config
 // to be sent down the subscribed channel when a new configuration
 // is loaded.
-func (ca *Agent) Subscribe(subscription chan<- Delta) {
-	ca.Lock()
-	defer ca.Unlock()
+func (ca *Agent) Subscribe(subscription DeltaChan) {
+	ca.mut.Lock()
+	defer ca.mut.Unlock()
 	ca.subscriptions = append(ca.subscriptions, subscription)
 }
 
+// Getter returns the current Config in a thread-safe manner.
+type Getter func() *Config
+
 // Config returns the latest config. Do not modify the config.
 func (ca *Agent) Config() *Config {
-	ca.Lock()
-	defer ca.Unlock()
+	ca.mut.RLock()
+	defer ca.mut.RUnlock()
 	return ca.c
 }
 
 // Set sets the config. Useful for testing.
 func (ca *Agent) Set(c *Config) {
-	ca.Lock()
-	defer ca.Unlock()
+	ca.mut.Lock()
+	defer ca.mut.Unlock()
 	var oldConfig Config
 	if ca.c != nil {
 		oldConfig = *ca.c
 	}
 	delta := Delta{oldConfig, *c}
+	ca.c = c
 	for _, subscription := range ca.subscriptions {
-		// we can't let unbuffered channels for subscriptions lock us up
-		// here, so we will send events best-effort into the channels we have
-		go func(out chan<- Delta) { out <- delta }(subscription)
+		go func(sub DeltaChan) { // wait a minute to send each event
+			end := time.NewTimer(time.Minute)
```

Comment:
Why do we need to try to send the event for a minute? It doesn't seem unreasonable to expect the consumer to be watching the channel when we send the event. That seems like more idiomatic go to me.

Reply:
Agreed with Cole here, not sure we needed the change

Reply:
* Previously we waited forever, which seems leaky to me
* Requiring a [non-blocking send](https://gobyexample.com/non-blocking-channel-operations) seems overly aggressive to me
* It seems most likely that within a minute the receiving thread will have either caught up or is dead.
  - Aka if it hasn't received the event within a minute, I think we can assume it will never receive the event.

If we want to switch this to something like:
```golang
for _, subscription := range ca.subscriptions {
  select {
  case sub <- delta:
  default:
  }
}
```
AKA getting rid of all the go routines, then that seems like the best solution... so as we can prove that this is not too aggressive.